File: backend/scripts/hardware_simulator.py
# Hardware Integration Script (Python example)
# This script simulates a sensor device sending readings to the backend

import requests
import time
import random

# Configuration
API_BASE = "http://localhost:3000"
EMAIL = "sensor@example.com"
PASSWORD = "sensor123"

def register_and_login():
    """Register or login to get access token"""
    # Try login first
    response = requests.post(f"{API_BASE}/auth/login", json={
        "email": EMAIL,
        "password": PASSWORD
    })
    
    if response.status_code != 201:
        # Register if login fails
        response = requests.post(f"{API_BASE}/auth/register", json={
            "email": EMAIL,
            "password": PASSWORD
        })
    
    return response.json()["access_token"]

def create_device(token, name, device_type):
    """Create a new device"""
    headers = {"Authorization": f"Bearer {token}"}
    response = requests.post(f"{API_BASE}/sensors/devices", 
        headers=headers,
        json={"name": name, "type": device_type}
    )
    return response.json()["id"]

def send_reading(token, device_id, value, unit):
    """Send a sensor reading"""
    headers = {"Authorization": f"Bearer {token}"}
    response = requests.post(
        f"{API_BASE}/sensors/devices/{device_id}/readings",
        headers=headers,
        json={"value": value, "unit": unit}
    )
    return response.json()

def simulate_temperature_sensor():
    """Simulate a temperature sensor sending data every 5 seconds"""
    print("Authenticating...")
    token = register_and_login()
    
    print("Creating temperature device...")
    device_id = create_device(token, "Classroom Temp Sensor", "temperature")
    
    print(f"Device created: {device_id}")
    print("Starting to send readings (Ctrl+C to stop)...")
    
    try:
        while True:
            # Simulate temperature reading (20-30°C with some variation)
            temp = 25 + random.uniform(-5, 5)
            reading = send_reading(token, device_id, round(temp, 2), "°C")
            print(f"Sent: {reading['value']}°C at {reading['timestamp']}")
            time.sleep(5)
    except KeyboardInterrupt:
        print("\nStopped.")

if __name__ == "__main__":
    simulate_temperature_sensor()

File: backend/scripts/hardware_simulator.js
// Hardware Integration Script (Node.js example)
// This script simulates an Arduino/ESP32 sending sensor data

const axios = require('axios');

const API_BASE = 'http://localhost:3000';
const EMAIL = 'sensor@example.com';
const PASSWORD = 'sensor123';

async function authenticate() {
  try {
    const response = await axios.post(`${API_BASE}/auth/login`, {
      email: EMAIL,
      password: PASSWORD,
    });
    return response.data.access_token;
  } catch {
    const response = await axios.post(`${API_BASE}/auth/register`, {
      email: EMAIL,
      password: PASSWORD,
    });
    return response.data.access_token;
  }
}

async function createDevice(token, name, type) {
  const response = await axios.post(
    `${API_BASE}/sensors/devices`,
    { name, type },
    { headers: { Authorization: `Bearer ${token}` } }
  );
  return response.data.id;
}

async function sendReading(token, deviceId, value, unit) {
  const response = await axios.post(
    `${API_BASE}/sensors/devices/${deviceId}/readings`,
    { value, unit },
    { headers: { Authorization: `Bearer ${token}` } }
  );
  return response.data;
}

async function simulateMultipleSensors() {
  console.log('Authenticating...');
  const token = await authenticate();

  console.log('Creating devices...');
  const tempDevice = await createDevice(token, 'Temperature Sensor', 'temperature');
  const humidDevice = await createDevice(token, 'Humidity Sensor', 'humidity');
  const lightDevice = await createDevice(token, 'Light Sensor', 'light');

  console.log(`Devices created: ${tempDevice}, ${humidDevice}, ${lightDevice}`);
  console.log('Starting to send readings (Ctrl+C to stop)...');

  setInterval(async () => {
    try {
      const temp = 25 + Math.random() * 10 - 5;
      const humidity = 50 + Math.random() * 20 - 10;
      const light = 300 + Math.random() * 400;

      await sendReading(token, tempDevice, parseFloat(temp.toFixed(2)), '°C');
      await sendReading(token, humidDevice, parseFloat(humidity.toFixed(2)), '%');
      await sendReading(token, lightDevice, parseFloat(light.toFixed(2)), 'lux');

      console.log(`Sent: ${temp.toFixed(2)}°C, ${humidity.toFixed(2)}%, ${light.toFixed(2)} lux`);
    } catch (err) {
      console.error('Error sending readings:', err.message);
    }
  }, 5000);
}

simulateMultipleSensors().catch(console.error);
